cpp/src/arrow/io/memory.h

```diff
@@ -65,14 +107,16 @@ class ARROW_EXPORT BufferOutputStream : public OutputStream {
   int64_t capacity_;
   int64_t position_;
   uint8_t* mutable_data_;
+  std::unique_ptr<Memcopy> memcopy_;
 };
 
 /// \brief Enables random writes into a fixed-size mutable buffer
 ///
 class ARROW_EXPORT FixedSizeBufferWriter : public WriteableFile {
  public:
   /// Input buffer must be mutable, will abort if not
-  explicit FixedSizeBufferWriter(const std::shared_ptr<Buffer>& buffer);
+  explicit FixedSizeBufferWriter(const std::shared_ptr<Buffer>& buffer,
+    std::unique_ptr<Memcopy> memcopy = nullptr);
```

Comment:
We could add other parameters, too, like a size threshold for parallelization (i.e. you wouldn't want to spawn a thread pool to copy 1024 bytes)